### ISSUE TITLE
Transitioning from screen with header to screen without header causes layout jump

### ISSUE THREAD

**Issue by yordis** [react-navigation/react-navigation#3517]
### Current Behavior

![screenflow](https://user-images.githubusercontent.com/4237280/36186921-f2ed890e-10f6-11e8-9c14-45f59cf49630.gif)

### Expected Behavior

The content shouldn't move up and down when the header is hidden. The content is jumping when you drag back.

### How to reproduce

https://snack.expo.io/HJuM4Vbvz

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 1.0.3
| react-native     | snack.expo version (no sure)
| node             | snack.expo version (no sure)
| npm or yarn      | snack.expo version (no sure)

**Comment by brentvatne** [comment]
You might want to use `headerMode: 'screen'` instead

**Comment by yordis** [comment]
@brentvatne but if I do that it will behave as Android and this was iOS, I am letting RN to decide which one should use.

**Comment by obibring** [comment]
I'm seeing a similar issue navigating forwards. Both the origin screen and the destination screen are nested inside the same parent StackNavigator (`headerMode: "none"` and `mode: "card"`), and this navigator is nested inside a grandparent StackNavigator with `headerMode: "screen"` and `mode: "modal"`.

**Comment by yordis** [comment]
@obibring well I am fine understanding that if I do not add the necessary margin top the content will have a bar on top of the image on this case, so that is why I pushed down as much as I could that look nice, but I do not expect all my layout to be moved.

**Comment by brentvatne** [comment]
fyi this is not a priority for us to fix right now but a pull request to make this transition smoother would be welcome

**Comment by yordis** [comment]
@brentvatne could you share at least where to look up for this? I have no clue about the internals and any help it is appreciated.

**Comment by brentvatne** [comment]
CardStack.js, Header.js

**Comment by danleechina** [comment]
I have noticed this too. And I also noticed that transitioning from screen without tab bar to tab bar causes layout jump

**Comment by vonovak** [comment]
@yordis just curious, what is the correct behavior in this case? how does ios handle such situations?

**Comment by yordis** [comment]
@vonovak you can do multiple effects but mainly they fade out the header while you move the card left or right 

But the header do not move the content (it is kind of overlapping the screens like floating rather than pushing the screens downs). Probably you can check Health from Apple itself and click in Sources.

In that case, they just transform the header into another header and the content itself fade out and the other one appear.

When you have this full screen you would disappear the header as it moves.

I am not quite sure how to tackle this to be honest, even YouTube has a unintended UI because of this. Normally Apple products do not have this kind of UI so they do not face this situation (as far as I know) 

Maybe the fix is to overlap the content and add a margin top only of your screen has a header. So when you swipe back you fade out the header and the screen already knows that it doesn’t have any header so it wouldn’t move. 🤔🤔🤔 not sure

**Comment by jiasaichao** [comment]
@yordis 
可以采用嵌套的方式，有头部的为子项。
无头部的：
```
const Index = StackNavigator({
    /**
     * 登录页
     */
    Login: {
        screen: LoginScreen,
        navigationOptions: {
            header: null
        }
    },
    /**
     * 主页
     */
    Home: {
        screen: HomeScreen,
        navigationOptions: {
            header: null
        }
    },
    Content: {
        screen: Content,
        navigationOptions: {
            header: null
        }
    }

});
```
有头部的
```
const Content = StackNavigator({
    AllList: {
        screen: AllListScreen,
        navigationOptions: ...
    },
    DeviceList: {
        screen: DeviceListScreen,
        navigationOptions: ...
    }
})@yordis 
可以采用嵌套的方式，有头部的为子项。
无头部的：
const Index = StackNavigator({
    /**
     * 登录页
     */
    Login: {
        screen: LoginScreen,
        navigationOptions: {
            header: null
        }
    },
    /**
     * 主页
     */
    Home: {
        screen: HomeScreen,
        navigationOptions: {
            header: null
        }
    },
    Content: {
        screen: Content,
        navigationOptions: {
            header: null
        }
    }

});
有头部的
const Content = StackNavigator({
    AllList: {
        screen: AllListScreen,
        navigationOptions: ...
    },
    DeviceList: {
        screen: DeviceListScreen,
        navigationOptions: ...
    }
})
```

**Comment by yordis** [comment]
@jiasaichao mate I understand nothing from your comment ... sorry.

**Comment by jiasaichao** [comment]
@yordis 
https://snack.expo.io/HkRKMHIOM

**Comment by danleechina** [comment]
This is just a workaround. But not fixing the bug.

**Comment by brentvatne** [comment]
I unlocked this issue at the request of @yordis

**Comment by yordis** [comment]
@jiasaichao thank you very much for the demo

@danleechina now watching how the demo works,

I am not sure if we are the ones that should change our architecture/code around the navigation of if this is an actual issue.

The downside of doing @jiasaichao is that it will require more navigation steps for avoiding this situation.

@brentvatne what is your opinion about it?

**Comment by terrysahaidak** [comment]
@jiasaichao did the right thing.

My solution for that issues will be next:
1. All the screen headers should be with position absolute.
2. Add top margin with the value of header's height to the Card, so we won't cover the Card's content by the header.
3. Add the ability to disable that margin-top. So header will cover the Card content. But it's ok if we can add an ability to make the header transparent.
4. What will cause the header transition behavior to be from transparent to non-transparent when we navigate from the screen without header, and from non-transparent to transparent in a reverse way.
5. For potential problems like unreached controls under the transparent header, we can apply tiny hack I'm using in some projects to animate the header (to animate move an ios search-box within the header): just move the header all over the viewport just after the transition has ended (transparent header won't be visible to the user even in modal screen) and the same thing but in reverse (move the header to the screen right before transition).

I'm not sure if that approach won't break any stuff. 

Is it possible to do something like that @brentvatne?

**Comment by terrysahaidak** [comment]
It will also allow us to make iOS 11 like transitions as well by setting a bigger margin for the first screen:

![ios11-navigator-animation](https://user-images.githubusercontent.com/7809008/36986334-3c0876fc-20a2-11e8-9652-f1042a114ad8.gif)

**Comment by brentvatne** [comment]
hi @terrysahaidak! that makes sense, i think we do need to absolute position the headers.

> It will also allow us to make iOS 11 like transitions as well just by setting a bigger margin for the screen with a big header

unfortunately margin cannot be animated with native driver, only properties that don't impact layout can (for now).

**Comment by yordis** [comment]
@brentvatne @terrysahaidak  I will try to target for that I think that is a good direction.

**Comment by brentvatne** [comment]
this is resolved in 2.0

**Comment by brentvatne** [comment]
here's the snack from the original report, updated to 2.0 https://snack.expo.io/Hya-5UATz
